Docs, part 18 (#507)
yiisoft/validator: docs/guide/rules.md

# Rules

## `Required` - requiring values

Use `Yiisoft\Validator\Rule\Required` rule to make sure value is provided. What values are considered empty can be
customized via `$emptyCallback` option. Normalization is not performed here, so only a callable or special class is

> needed. For more details see "Skipping empty values" section.

Reviewer: How about linking to the section?

$rule = new Nested([
    'title' => [new Required()],
    'author' => new Nested([
        'name' => [new HasLength(min: 3)],
        'age' => [new Number(min: 18)],
    )];
]);
$errors = $validator->validate($data, [$rule])->getErrorMessagesIndexedByPath();
```

### Other configuration options

A dot notation can be used as an alternative way of configuration. In this case the example above will be presented as
following:

```php
use Yiisoft\Validator\Rule\HasLength;
use Yiisoft\Validator\Rule\Nested;
use Yiisoft\Validator\Rule\Number;
use Yiisoft\Validator\Rule\Required;

$rule = new Nested([
    'title' => [new Required()],
    'author.name' => [new HasLength(min: 3)],
    'author.age' => [new Number(min: 18)],
)];
```

It's also possible to combine both of these approaches:

```php
use Yiisoft\Validator\Rule\HasLength;
use Yiisoft\Validator\Rule\Nested;
use Yiisoft\Validator\Rule\Number;
use Yiisoft\Validator\Rule\Required;

$data = ['author' => ['name' => 'Alexey', 'age' => '31']];
$rule = new Nested([
    'author' => new Nested([
        'name' => [new HasLength(min: 3)],
        'age' => [new Number(min: 18)],
    )];
]);
```

### Advanced usage

> A more complex real-life example is a chart that is made of points. This data is represented as arrays. `Nested` can be

Reviewer: Would be cool to add a line chart image here but that could be done later on the "enhancements" stage.

Author: Added to TODO list (optional).

Author: Decided to add it in this PR.

Author: Done, please check.

Reviewer: 👍

    // ...
    'charts.0.points.0.rgb.0' => ['Value must be no less than 0. -1 given.'],
    // ...
];
```

#### Using shortcut

A shortcut can be used to simplify `Nested` and `Each` combinations:

```php
use Yiisoft\Validator\Rule\Count;
use Yiisoft\Validator\Rule\Each;
use Yiisoft\Validator\Rule\Nested;

$rule = new Nested([
    'charts.*.points.*.coordinates.x' => [new Number(min: -10, max: 10)],
    'charts.*.points.*.coordinates.y' => [new Number(min: -10, max: 10)],
    'charts.*.points.*.rgb' => [
        new Count(exactly: 3);
        new Number(min: 0, max: 255),
    ]),
]);
```

With additional grouping it can also be rewritten like this:

```php
use Yiisoft\Validator\Rule\Count;
use Yiisoft\Validator\Rule\Each;
use Yiisoft\Validator\Rule\Nested;

$rule = new Nested([
    'charts.*.points.*.coordinates' => new Nested([
        'x' => [new Number(min: -10, max: 10)],
        'y' => [new Number(min: -10, max: 10)],
    ]),
    'charts.*.points.*.rgb' => [
        new Count(exactly: 3);
        new Number(min: 0, max: 255),
    ]),
]);
```

This is less verbose, but the downside of this approach is that you can't additionally configure dynamically generated

> `Nested` and `Each` pairs. If you need this, please refer to example provided in "Basic usage" section.

Reviewer: Would be great to have a section linked.

```php
use Yiisoft\Validator\Rule\Nested;

$rule = new Nested([
    'author\.data' => new Nested([
        'name\.surname' => [
            new HasLength(min: 3),
        ],
    ]),
]);
$data = [
    'author.data' => [
        'name.surname' => 'Dmitriy',
    ],
];
```

The same applies to shortcut:

```php
use Yiisoft\Validator\Rule\Nested;

$rule = new Nested([
    'charts\.list.*.points\*list.*.coordinates\.data.x' => [
        // ...
    ],
    'charts\.list.*.points\*list.*.coordinates\.data.y' => [
        // ...
    ],
    'charts\.list.*.points\*list.*.rgb' => [
        // ...
    ],
]);
$data = [
    'charts.list' => [
        [
            'points*list' => [
                [
                    'coordinates.data' => ['x' => -11, 'y' => 11], 'rgb' => [-1, 256, 0],
                ],
            ],
        ],
    ],
];
```

## `StopOnError` - lazy or early exiting validation

When validation propagation is not needed `\Yiisoft\Validator\Rule\StopOnError` rule may be used.

```php
use Yiisoft\Validator\Rule\StopOnError;

new StopOnError([
    new Rule1();
    new Rule2();
    new Rule3();
])
```

> When the validator get negative result of validation it stop all the rest rules inside `StopOnError`.

Reviewer: ```suggestion
When the validator gets a negative result of validation it stops the rest of the rules defined inside `StopOnError`.
```

> According to the example above, if `Rule2` fails `Rule3` won't even be tried to run.

Reviewer: ```suggestion
According to the example above, if `Rule2` fails `Rule3` won't be run.
```
